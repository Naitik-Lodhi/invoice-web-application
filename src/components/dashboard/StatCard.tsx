import { Card, CardContent, Typography, Box } from "@mui/material";
import TrendingUpIcon from '@mui/icons-material/TrendingUp';
import TrendingDownIcon from '@mui/icons-material/TrendingDown';

interface StatCardProps {
  title: string;
  value: string | number;
  subtext: string;
  trend?: number; // optional: percentage change
  icon?: React.ReactNode; // optional: icon for the card
}

const StatCard = ({ title, value, subtext, trend, icon }: StatCardProps) => {
  return (
    <Card
      sx={{
        height: { xs: "160px", sm: "180px", md: "214px" },
        boxShadow: "0 2px 8px rgba(0, 0, 0, 0.08)",
        cursor: "pointer",
      }}
    >
      <CardContent
        sx={{
          height: "100%",
          display: "flex",
          flexDirection: "column",
          justifyContent: "space-between",
          p: { xs: 2, sm: 2.5, md: 3 },
        }}
      >
        {/* Top Section: Title with optional icon */}
        <Box sx={{ display: "flex", justifyContent: "space-between", alignItems: "flex-start" }}>
          <Typography
            color="#737373"
            sx={{
              fontSize: { xs: "0.75rem", sm: "0.8rem", md: "0.85rem" },
              fontWeight: 500,
              lineHeight: 1.2,
              maxWidth: "80%",
            }}
          >
            {title}
          </Typography>
          {icon && (
            <Box sx={{ color: "#A3A3A3", opacity: 0.6 }}>
              {icon}
            </Box>
          )}
        </Box>

        {/* Middle Section: Main Value */}
        <Box sx={{ my: { xs: 1, sm: 1.5 } }}>
          <Typography
            variant="h4"
            component="div"
            sx={{
              fontWeight: 600,
              color: "#171717",
              fontSize: { xs: "1.5rem", sm: "1.75rem", md: "2.125rem", lg: "2.25rem" },
              lineHeight: 1,
              letterSpacing: "-0.02em",
            }}
          >
            {value}
          </Typography>
        </Box>

        {/* Bottom Section: Subtext with optional trend */}
        <Box sx={{ display: "flex", alignItems: "center", gap: 1 }}>
          <Typography
            variant="caption"
            sx={{
              color: "#A3A3A3",
              fontSize: { xs: "0.7rem", sm: "0.75rem" },
            }}
          >
            {subtext}
          </Typography>
          {trend !== undefined && (
            <Box
              sx={{
                display: "flex",
                alignItems: "center",
                gap: 0.5,
                color: trend >= 0 ? "#10b981" : "#ef4444",
              }}
            >
              {trend >= 0 ? (
                <TrendingUpIcon sx={{ fontSize: "1rem" }} />
              ) : (
                <TrendingDownIcon sx={{ fontSize: "1rem" }} />
              )}
              <Typography sx={{ fontSize: "0.75rem", fontWeight: 600 }}>
                {Math.abs(trend)}%
              </Typography>
            </Box>
          )}
        </Box>
      </CardContent>
    </Card>
  );
};

export default StatCard;