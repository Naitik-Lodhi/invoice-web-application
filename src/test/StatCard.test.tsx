import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import StatCard from '../components/dashboard/StatCard';

describe('StatCard Component', () => {
  
  // ============================================
  // UI-01: Basic Rendering
  // ============================================
  describe('UI-01: Basic Rendering', () => {
    it('should render card with title, value, and subtext', () => {
      render(
        <StatCard 
          title="Total Invoices" 
          value="125" 
          subtext="This Month" 
        />
      );

      expect(screen.getByText('Total Invoices')).toBeInTheDocument();
      expect(screen.getByText('125')).toBeInTheDocument();
      expect(screen.getByText('This Month')).toBeInTheDocument();
    });

    it('should render with optional icon', () => {
      const TestIcon = () => <span data-testid="test-icon">ðŸ“Š</span>;
      
      render(
        <StatCard 
          title="Revenue" 
          value="$50,000" 
          subtext="Today"
          icon={<TestIcon />}
        />
      );

      expect(screen.getByTestId('test-icon')).toBeInTheDocument();
    });

    it('should render with optional trend indicator', () => {
      render(
        <StatCard 
          title="Sales" 
          value="$1,234" 
          subtext="Week"
          trend={15.5}
        />
      );

      // Check for trend percentage
      expect(screen.getByText('15.5%')).toBeInTheDocument();
    });
  });

  // ============================================
  // UI-02: Correct Values Display
  // ============================================
  describe('UI-02: Card Values', () => {
    it('should display invoice count correctly', () => {
      render(
        <StatCard 
          title="Invoices" 
          value={250} 
          subtext="Total" 
        />
      );

      expect(screen.getByText('250')).toBeInTheDocument();
    });

    it('should display formatted numbers', () => {
      render(
        <StatCard 
          title="Total Amount" 
          value="1,234" 
          subtext="This Month" 
        />
      );

      expect(screen.getByText('1,234')).toBeInTheDocument();
    });

    it('should handle string and number values', () => {
      const { rerender } = render(
        <StatCard title="Count" value={100} subtext="Items" />
      );
      expect(screen.getByText('100')).toBeInTheDocument();

      rerender(
        <StatCard title="Count" value="100" subtext="Items" />
      );
      expect(screen.getByText('100')).toBeInTheDocument();
    });
  });

  // ============================================
  // UI-03: Currency Symbol Display
  // ============================================
  describe('UI-03: Currency Symbol', () => {
    it('should display currency symbol with amount', () => {
      render(
        <StatCard 
          title="Total Amount" 
          value="$12,345.67" 
          subtext="This Month" 
        />
      );

      expect(screen.getByText('$12,345.67')).toBeInTheDocument();
    });

    it('should work with different currency symbols', () => {
      const { rerender } = render(
        <StatCard title="Revenue" value="â‚¹10,000" subtext="Today" />
      );
      expect(screen.getByText('â‚¹10,000')).toBeInTheDocument();

      rerender(
        <StatCard title="Revenue" value="â‚¬5,000" subtext="Today" />
      );
      expect(screen.getByText('â‚¬5,000')).toBeInTheDocument();

      rerender(
        <StatCard title="Revenue" value="Â£3,000" subtext="Today" />
      );
      expect(screen.getByText('Â£3,000')).toBeInTheDocument();
    });
  });

  // ============================================
  // UI-04: Thousand Separator Formatting
  // ============================================
  describe('UI-04: Thousand Separator', () => {
    it('should display numbers with thousand separator', () => {
      render(
        <StatCard 
          title="Total" 
          value="12,345.67" 
          subtext="Amount" 
        />
      );

      expect(screen.getByText('12,345.67')).toBeInTheDocument();
    });

    it('should format large numbers correctly', () => {
      const testCases = [
        { value: '1,234', expected: '1,234' },
        { value: '12,345', expected: '12,345' },
        { value: '123,456', expected: '123,456' },
        { value: '1,234,567', expected: '1,234,567' },
        { value: '12,345,678.90', expected: '12,345,678.90' },
      ];

      testCases.forEach(({ value, expected }) => {
        const { unmount } = render(
          <StatCard title="Amount" value={value} subtext="Total" />
        );
        expect(screen.getByText(expected)).toBeInTheDocument();
        unmount();
      });
    });

    it('should handle decimal places correctly', () => {
      render(
        <StatCard 
          title="Amount" 
          value="1,234.56" 
          subtext="Total" 
        />
      );

      expect(screen.getByText('1,234.56')).toBeInTheDocument();
    });
  });

  // ============================================
  // UI-05: Zero Data Handling
  // ============================================
  describe('UI-05: Zero Data', () => {
    it('should display zero when no invoices', () => {
      render(
        <StatCard 
          title="Invoices" 
          value={0} 
          subtext="This Month" 
        />
      );

      expect(screen.getByText('0')).toBeInTheDocument();
      expect(screen.getByText('This Month')).toBeInTheDocument();
    });

    it('should display zero amount with currency', () => {
      render(
        <StatCard 
          title="Total Amount" 
          value="$0.00" 
          subtext="Today" 
        />
      );

      expect(screen.getByText('$0.00')).toBeInTheDocument();
    });

    it('should handle empty string as zero', () => {
      render(
        <StatCard 
          title="Count" 
          value="" 
          subtext="Items" 
        />
      );

      // Empty string will be rendered as is
      expect(screen.getByText('Count')).toBeInTheDocument();
    });
  });

  // ============================================
  // UI-06: Mobile Responsive Layout
  // ============================================
  describe('UI-06: Mobile Responsiveness', () => {
    it('should render correctly on mobile viewport', () => {
      // Mock mobile viewport
      global.innerWidth = 375;
      global.innerHeight = 667;

      render(
        <StatCard 
          title="Invoices" 
          value="125" 
          subtext="This Month" 
        />
      );

      // Card should still render all elements
      expect(screen.getByText('Invoices')).toBeInTheDocument();
      expect(screen.getByText('125')).toBeInTheDocument();
      expect(screen.getByText('This Month')).toBeInTheDocument();
    });

    it('should render correctly on tablet viewport', () => {
      global.innerWidth = 768;
      global.innerHeight = 1024;

      render(
        <StatCard 
          title="Revenue" 
          value="$50,000" 
          subtext="Today" 
        />
      );

      expect(screen.getByText('Revenue')).toBeInTheDocument();
    });

    it('should render correctly on desktop viewport', () => {
      global.innerWidth = 1920;
      global.innerHeight = 1080;

      render(
        <StatCard 
          title="Total" 
          value="$100,000" 
          subtext="Year" 
        />
      );

      expect(screen.getByText('Total')).toBeInTheDocument();
    });
  });

  // ============================================
  // Additional Edge Cases
  // ============================================
  describe('Edge Cases', () => {
    it('should handle very long titles', () => {
      render(
        <StatCard 
          title="This is a very long title that might wrap" 
          value="100" 
          subtext="Data" 
        />
      );

      expect(screen.getByText('This is a very long title that might wrap')).toBeInTheDocument();
    });

    it('should handle very long values', () => {
      render(
        <StatCard 
          title="Amount" 
          value="$999,999,999.99" 
          subtext="Maximum" 
        />
      );

      expect(screen.getByText('$999,999,999.99')).toBeInTheDocument();
    });

    it('should handle negative values', () => {
      render(
        <StatCard 
          title="Balance" 
          value="-$1,000" 
          subtext="Deficit" 
        />
      );

      expect(screen.getByText('-$1,000')).toBeInTheDocument();
    });

    it('should handle trend with negative value', () => {
      render(
        <StatCard 
          title="Sales" 
          value="$500" 
          subtext="Down"
          trend={-12.5}
        />
      );

      expect(screen.getByText('12.5%')).toBeInTheDocument();
    });
  });

  // ============================================
  // Accessibility Tests
  // ============================================
  describe('Accessibility (A11y)', () => {
    it('should have proper semantic structure', () => {
      const { container } = render(
        <StatCard 
          title="Invoices" 
          value="125" 
          subtext="This Month" 
        />
      );

      // Check if card is rendered
      expect(container.querySelector('.MuiCard-root')).toBeInTheDocument();
    });

    it('should be keyboard accessible', () => {
      const { container } = render(
        <StatCard 
          title="Revenue" 
          value="$1,000" 
          subtext="Today" 
        />
      );

      const card = container.querySelector('.MuiCard-root');
      expect(card).toBeInTheDocument();
    });
  });

  // ============================================
  // Snapshot Tests
  // ============================================
  describe('Snapshot Tests', () => {
    it('should match snapshot for basic card', () => {
      const { container } = render(
        <StatCard 
          title="Invoices" 
          value="125" 
          subtext="This Month" 
        />
      );

      expect(container.firstChild).toMatchSnapshot();
    });

    it('should match snapshot for card with trend', () => {
      const { container } = render(
        <StatCard 
          title="Sales" 
          value="$1,000" 
          subtext="Week"
          trend={15}
        />
      );

      expect(container.firstChild).toMatchSnapshot();
    });
  });
});