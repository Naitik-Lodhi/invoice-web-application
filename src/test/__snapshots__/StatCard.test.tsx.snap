// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`StatCard Component > Snapshot Tests > should match snapshot for basic card 1`] = `
<div
  class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-1c9z231-MuiPaper-root-MuiCard-root"
  style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);"
>
  <div
    class="MuiCardContent-root css-1p1pav1-MuiCardContent-root"
  >
    <div
      class="MuiBox-root css-1qdyvok"
    >
      <p
        class="MuiTypography-root MuiTypography-body1 css-11cn1t9-MuiTypography-root"
      >
        Invoices
      </p>
    </div>
    <div
      class="MuiBox-root css-i30p6k"
    >
      <div
        class="MuiTypography-root MuiTypography-h4 css-ih48ct-MuiTypography-root"
      >
        125
      </div>
    </div>
    <div
      class="MuiBox-root css-axw7ok"
    >
      <span
        class="MuiTypography-root MuiTypography-caption css-1d4omqi-MuiTypography-root"
      >
        This Month
      </span>
    </div>
  </div>
</div>
`;

exports[`StatCard Component > Snapshot Tests > should match snapshot for card with trend 1`] = `
<div
  class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-1c9z231-MuiPaper-root-MuiCard-root"
  style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);"
>
  <div
    class="MuiCardContent-root css-1p1pav1-MuiCardContent-root"
  >
    <div
      class="MuiBox-root css-1qdyvok"
    >
      <p
        class="MuiTypography-root MuiTypography-body1 css-11cn1t9-MuiTypography-root"
      >
        Sales
      </p>
    </div>
    <div
      class="MuiBox-root css-i30p6k"
    >
      <div
        class="MuiTypography-root MuiTypography-h4 css-ih48ct-MuiTypography-root"
      >
        $1,000
      </div>
    </div>
    <div
      class="MuiBox-root css-axw7ok"
    >
      <span
        class="MuiTypography-root MuiTypography-caption css-1d4omqi-MuiTypography-root"
      >
        Week
      </span>
      <div
        class="MuiBox-root css-ugu8rw"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1wdc28j-MuiSvgIcon-root"
          data-testid="TrendingUpIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="m16 6 2.29 2.29-4.88 4.88-4-4L2 16.59 3.41 18l6-6 4 4 6.3-6.29L22 12V6z"
          />
        </svg>
        <p
          class="MuiTypography-root MuiTypography-body1 css-1j6cjbn-MuiTypography-root"
        >
          15
          %
        </p>
      </div>
    </div>
  </div>
</div>
`;
